### PR TITLE
Tracegases as associative list

### PR DESCRIPTION
**Pull request by Azrun**
[performance][help wanted]
## About the PR 
Transitions trace_gases to also have an associative list allow for random access.  Original list maintained for benefits of sequential access on non-associative list.
* Hits and Misses for known types will improve
* Iteration over list will see minor reduction in performance

Favored associative list instead of `locate()` as per:
* https://discord.com/channels/182249960895545344/517361638504923136/789362743589339187 (hash miss)
* https://discord.com/channels/182249960895545344/517361638504923136/788565364224163871 (hit)

~``as()` removed in for loop per~
* ~https://discord.com/channels/182249960895545344/517361638504923136/794499908161896479~

Sequential access of trace_gases converted to use typecheckless iteration.

Fix for overlays being stuck on simulated turf converting to other types.

- Overlay determination reworked to use bitfield instead of string generation and parameter parsing.... 😕 

### Initial Profiling Findings
d7fb6ad documents scenario used for profiling, 3 runs pre and post proposed implementation were performed with said scenario and summed to generate data in the table below.
[tracegas_map.zip](https://github.com/goonstation/goonstation/files/5759899/tracegas_map.zip)

Preliminary analysis seems to show general improvements with the exception of heat capacity and any other iteration instead of hashing operation on the list.  Subsequent analysis showed that improvements appeared to be a bit of a wash, as the only improvements WERE the hashing operations and everything else slowed down. 

### 

## Why's this needed? 
Minor improvement to ATMOS performance when trace gases are present.

### CODE REVIEW

**Comment by Azrun** [comment]
@pali6 or others...
This seem worth pursuing? 😛
If so, I plan on setting up something automated to get a larger data set over the weekend. 

Or should trace_gases just be a datum we stick scalars on? I like the list (current or this implementation) because it allows for `secrets`. 🤷

**Comment by Azrun** [comment]
Woohooo inconsistent data! 😭

**Comment by ZeWaka** [comment]
You should also be looking into overtime, because that is what we humans perceive as 'lag.'

**Comment by Azrun** [comment]
* Updating scenario to remove more doors to allow gas/fire/traces to spread more.
* Disabled lagcheck to allow for tick overruns to occur to see profile un-massaged _work_, as lag checks would spread out cpu load across ticks which is fine since ATMOS is running at 0.5Hz. 
* Forced hard-reset as behavior of the server restarting correctly or not was inconsistent.

Preliminary 2 pre and 2 post change profiles are more representative of expected results. 🎉 Will run a larger sample overnight.

name | selfΔ | totalΔ | overtimeΔ | self% | total% | overtime% | over_pre | over_post | calls_pre | calls_post
---- | ----- | ------ | --------- | ----- | ------ | --------- | -------- | --------- | --------- | ----------
/datum/gas_mixture/proc/heat_capacity_full |  8.926434e-08 |  8.944301e-08 |     3.088 |   5.050571 |   5.023592 |   9.921285  |  31.125  |   34.213 | 22837927.0 | 23587863.0
/datum/gas_mixture/proc/share | -2.939823e-07 | -6.097619e-07 |    -5.871 |  -2.364444 |  -3.688467 |  -4.291165  | 136.816  |  130.945 | 15361445.0 | 16864968.0
/datum/gas_mixture/proc/heat_capacity_archived... |  5.026002e-08 |  4.908096e-08 |     0.594 |   3.041605 |   2.946842 |   3.562646  |  16.673  |   17.267 | 12951324.0 | 13057666.0
/datum/gas_mixture/proc/react | -1.337562e-07 | -2.734772e-07 |    -0.173 | -10.372563 |  -8.620499 |  -1.709148  |  10.122  |    9.949 | 12326295.0 | 13603995.0
/datum/gas_mixture/proc/check_tile_graphic | -3.189151e-07 | -3.194586e-07 |    -2.625 | -20.287671 | -20.130668 | -21.474149  |  12.224  |    9.599 | 11698097.0 | 12902119.0
/turf/simulated/return_air | -7.798557e-08 | -8.266853e-08 |    -0.045 | -11.476623 | -11.768426 |  -4.193849  |   1.073  |    1.028 |  2615094.0 |  2792873.0
/turf/proc/hotspot_expose | -1.261525e-07 | -1.716812e-07 |    -0.105 | -10.014938 |  -8.284053 |  -4.790146  |   2.192  |    2.087 |  2557867.0 |  2730503.0
/turf/simulated/hotspot_expose | -4.338345e-07 | -1.277621e-06 |     0.461 |  -5.951574 |  -7.567464 |   3.639665  |  12.666  |   13.127 |  2556998.0 |  2729750.0
/datum/gas_mixture/proc/fire | -7.257672e-07 | -6.473894e-07 |     0.062 | -13.540064 |  -7.537285 |   0.686372  |   9.033  |    9.095 |  2475679.0 |  2728737.0
/turf/simulated/super_conduct | -7.348351e-09 |  2.416187e-07 |     1.680 |  -0.105492 |   1.526758 |  10.292226  |  16.323  |   18.003 |  2347336.0 |  2590243.0
/datum/gas_mixture/proc/mimic | -5.819255e-07 | -8.762113e-07 |    -1.829 | -10.640502 | -13.500936 | -23.508997  |   7.780  |    5.951 |  2204621.0 |  2212586.0

**Review by pali6 — APPROVED** [review]
Sensible optimization.

**Comment by pali6** [comment]
> Or should trace_gases just be a datum we stick scalars on? I like the list (current or this implementation) because it allows for secrets. 🤷
I like the datum solution too (as I was one of the people to suggest it originally). I don't think it disallows secrets. Though, I also don't think there's any pressing need to implement it now.

**Comment by Azrun** [comment]
> I like the datum solution too (as I was one of the people to suggest it originally).
All I could remember was "pali wasn't the biggest fan of the associative list and suggested something about a datum" and then I couldn't seem to find conversation in discord. 😊 

There are some _nice_ optimizations as part of this change but overall due to the ubiquity of `heat_capacity_full`, `heat_capacity_archived_full`, and any other call checking the entire airgroup that it ends up being a wash due to the overhead of iterating over an associative list.  I think I'll try adding some helper functions for interacting with trace gases to abstract behavior and have the associated list an additional property of gas_mixtures to have the best of both worlds.

Start-Cycle 100
name | avg_self_pre | avg_self_post | selfΔ | totalΔ | overtimeΔ | self% | total% | overtime% | over_pre | over_post | calls_pre | calls_post
---- | ------------ | ------------- | ----- | ------ | --------- | ----- | ------ | --------- | -------- | --------- | --------- | ---------- 
/datum/controller/air_system/assemble_group_turf | 0.001771 |0.001750 | -0.000021 | -0.000343 | -0.001 |-1.209677|-9.448819| -16.666667 |0.006| 0.005|35.0| 28.0
/datum/controller/air_system/process | 0.000098 |0.000069 | -0.000029 |0.015130 | -0.008 | -30.029388| 3.477372| -23.529412 |0.034| 0.026| 388.0|394.0
/datum/controller/air_system/process_groups | 0.000817 |0.000782 | -0.000035 |0.013475 |0.007 |-4.251088| 3.337670| 4.794521 |0.146| 0.153| 388.0|395.0
/datum/controller/air_system/process_high_pres... | 0.000239 |0.000253 |0.000014 | -0.000022 |0.007 | 5.831067|-0.508256| 7.446809 |0.094| 0.101| 393.0|399.0
/datum/controller/air_system/process_rebuild_s... | 0.000500 |0.000500 |0.000000 | -0.001579 | -0.003 | 0.000000| -14.715719| -75.000000 |0.004| 0.001|30.0| 26.0
/datum/controller/air_system/process_singletons | 0.000140 |0.000148 |0.000008 |0.000142 |0.004 | 5.925080| 2.492364| 7.272727 |0.055| 0.059| 393.0|398.0
/datum/controller/air_system/process_super_con... | 0.000883 |0.000791 | -0.000091 |0.000644 | -0.033 | -10.362330| 2.668068|-9.510086 |0.347| 0.314| 393.0|398.0
/datum/controller/air_system/process_tiles_to_... | 0.000000 |0.000000 |0.000000 | -0.000019 |0.000 |NaN| -29.096447|NaN |0.000| 0.000| 388.0|394.0
/datum/controller/air_system/process_update_tiles | 0.000046 |0.000086 |0.000040 |0.000055 |0.002 |87.346939| 2.475625|inf |0.000| 0.002| 153.0|140.0

Cycle 100-200
name | avg_self_pre | avg_self_post | selfΔ | totalΔ | overtimeΔ | self% | total% | overtime% | over_pre | over_post | calls_pre | calls_post
---- | ------------ | ------------- | ----- | ------ | --------- | ----- | ------ | --------- | -------- | --------- | --------- | ---------- 
/datum/controller/air_system/assemble_group_turf | 0.001750 |0.001647 | -1.029412e-04 | -0.000497 |0.001 |-5.882353| -13.970588|16.666667 |0.006| 0.007|36.0| 34.0
/datum/controller/air_system/process | 0.000137 |0.000130 | -7.633472e-06 |0.013160 | -0.002 |-5.556167| 3.107598|-1.739130 |0.115| 0.113| 888.0|894.0
/datum/controller/air_system/process_groups | 0.000843 |0.000813 | -3.006065e-05 |0.011975 |0.023 |-3.563933| 3.077977| 8.070175 |0.285| 0.308| 888.0|895.0
/datum/controller/air_system/process_high_pres... | 0.000289 |0.000266 | -2.306283e-05 |0.000016 | -0.019 |-7.982599| 0.400684|-7.364341 |0.258| 0.239| 893.0|899.0
/datum/controller/air_system/process_rebuild_s... | 0.000484 |0.000469 | -1.512097e-05 | -0.002014 | -0.002 |-3.125000| -19.270833| -50.000000 |0.004| 0.002|31.0| 32.0
/datum/controller/air_system/process_singletons | 0.000132 |0.000135 |2.605017e-06 | -0.000016 |0.003 | 1.971424|-0.300276| 2.542373 |0.118| 0.121| 893.0|898.0
/datum/controller/air_system/process_super_con... | 0.000901 |0.000875 | -2.617737e-05 |0.000739 | -0.020 |-2.903900| 2.887278|-2.484472 |0.805| 0.785| 893.0|898.0
/datum/controller/air_system/process_tiles_to_... | 0.000001 |0.000001 | -7.557893e-09 | -0.000006 |0.000 |-0.671141| -19.772845|NaN |0.000| 0.000| 888.0|894.0
/datum/controller/air_system/process_update_tiles | 0.000039 |0.000065 |2.602171e-05 |0.000104 |0.004 |67.078189| 6.531804|inf |0.000| 0.004| 232.0|216.0

Cycle 200-500
name | avg_self_pre | avg_self_post | selfΔ | totalΔ | overtimeΔ | self% | total% | overtime% | over_pre | over_post | calls_pre | calls_post
---- | ------------ | ------------- | ----- | ------ | --------- | ----- | ------ | --------- | -------- | --------- | --------- | ---------- 
/datum/controller/air_system/assemble_group_turf | 2.112903e-03 |0.002463 |3.500597e-04 |0.000057 |0.004 | 16.567713 |1.307190 |40.000000 |0.010| 0.014|62.0| 54.0
/datum/controller/air_system/process | 1.260469e-04 |0.000118 | -7.834704e-06 |0.016221 | -0.017 | -6.215705 |4.280815 |-5.985915 |0.284| 0.267|2388.0| 2394.0
/datum/controller/air_system/process_groups | 8.835846e-04 |0.000860 | -2.387687e-05 |0.014491 |0.037 | -2.702273 |4.149099 | 5.901116 |0.627| 0.664|2388.0| 2395.0
/datum/controller/air_system/process_high_pres... | 1.608859e-04 |0.000142 | -1.916020e-05 |0.000018 | -0.045 |-11.909183 |0.796354 | -11.688312 |0.385| 0.340|2393.0| 2399.0
/datum/controller/air_system/process_rebuild_s... | 4.363636e-04 |0.000549 |1.126560e-04 | -0.000549 |0.000 | 25.816993 | -4.477221 | 0.000000 |0.004| 0.004|55.0| 51.0
/datum/controller/air_system/process_singletons | 1.249478e-04 |0.000136 |1.058184e-05 |0.000024 |0.026 |8.469014 |0.541735 | 8.695652 |0.299| 0.325|2393.0| 2398.0
/datum/controller/air_system/process_super_con... | 8.190556e-04 |0.000823 |3.713395e-06 |0.001595 |0.012 |0.453375 |7.163275 | 0.612245 |1.960| 1.972|2393.0| 2398.0
/datum/controller/air_system/process_tiles_to_... | 4.187605e-07 |0.000001 |8.343724e-07 |0.000002 |0.000 |199.248120 | 18.452381 |NaN |0.000| 0.000|2388.0| 2394.0
/datum/controller/air_system/process_update_tiles | 4.922280e-05 |0.000059 |9.600731e-06 |0.000010 |0.003 | 19.504644 |0.583752 | 300.000000 |0.001| 0.004| 386.0|357.0 

ALL	
name | avg_self_pre | avg_self_post | selfΔ | totalΔ | overtimeΔ | self% | total% | overtime% | over_pre | over_post | calls_pre | calls_post
---- | ------------ | ------------- | ----- | ------ | --------- | ----- | ------ | --------- | -------- | --------- | --------- | ---------- 
/datum/controller/air_system/assemble_group_turf | 1.984615e-03 |0.002195 |2.107869e-04 | -1.031830e-04 |0.013 |10.621046| -2.518234 |40.625000 |0.032| 0.045| 195.0|174.0
/datum/controller/air_system/process | 1.164640e-04 |0.000111 | -5.106709e-06 |8.729894e-03 | -0.033 |-4.384797|2.474099 |-3.518124 |0.938| 0.905|8552.0| 8576.0
/datum/controller/air_system/process_groups | 9.326473e-04 |0.000943 |1.059276e-05 |7.614972e-03 |0.092 | 1.135773|2.338818 | 4.362257 |2.109| 2.201|8552.0| 8580.0
/datum/controller/air_system/process_high_pres... | 1.525898e-04 |0.000145 | -7.289688e-06 | -1.106153e-05 | -0.059 |-4.777309| -0.509700 |-4.510703 |1.308| 1.249|8572.0| 8596.0
/datum/controller/air_system/process_rebuild_s... | 4.561404e-04 |0.000525 |6.855101e-05 | -9.015595e-04 | -0.004 |15.028490| -7.731528 | -25.000000 |0.016| 0.012| 171.0|162.0
/datum/controller/air_system/process_singletons | 1.231918e-04 |0.000130 |6.812868e-06 | -7.379164e-06 |0.061 | 5.530294| -0.166946 | 5.776515 |1.056| 1.117|8572.0| 8592.0
/datum/controller/air_system/process_super_con... | 7.451003e-04 |0.000729 | -1.651560e-05 |9.439680e-04 | -0.131 |-2.216560|4.656661 |-2.051041 |6.387| 6.256|8572.0| 8592.0
/datum/controller/air_system/process_tiles_to_... | 7.015903e-07 |0.000001 |3.478500e-07 |6.610132e-07 |0.000 |49.580224|4.790665 |NaN |0.000| 0.000|8552.0| 8576.0
/datum/controller/air_system/process_update_tiles | 3.843605e-05 |0.000051 |1.263365e-05 |4.565430e-05 |0.014 |32.869280|3.375421 | 700.000000 |0.002| 0.016|1509.0| 1449.0

**Comment by Azrun** [comment]
Redid PR to add associative list in addition to the existing list.  Abstracted access to some helper functions.  Added optimized helper functions inline in a couple of key places to eliminate unnecessarily decisions.  Cleaned up some of the sequential iterators to use typecheckless.

Revamped scenario to use a modified version of the stress test map.  Added some landmarks, modified some doors, adjusted some turf to show pipes.

name | avg_self_pre | avg_self_post | selfΔ | totalΔ | overtimeΔ | self% | total% | overtime% | over_pre | over_post | calls_pre | calls_post
---- | ------------ | ------------- | ----- | ------ | --------- | ----- | ------ | --------- | -------- | --------- | --------- | ---------- 
/datum/controller/air_system/process | 0.000236 | 0.000218 | -1.779615e-05 | -2.673081e-02 | -0.298 | -7.537888 | -15.835921 | -9.220297 | 3.232 | 2.934 | 17900.0 | 17788.0
/datum/controller/air_system/process_groups | 0.002696 | 0.002101 | -5.945069e-04 | -2.541170e-02 | -4.136 | -22.054366 | -16.384491 | -21.493530 | 19.243 | 15.107 | 17900.0 | 17788.0
/datum/controller/air_system/process_high_pres... | 0.000103 | 0.000093 | -9.466835e-06 | -5.386736e-05 | -0.181 | -9.235183 | -5.520912 | -9.869138 | 1.834 | 1.653 | 17940.0 | 17820.0
/datum/controller/air_system/process_singletons | 0.000009 | 0.000012 | 2.306604e-06 | -7.778266e-06 | 0.028 | 25.079074 | -0.939108 | 24.561404 | 0.114 | 0.142 | 17940.0 | 17820.0
/datum/controller/air_system/process_super_con... | 0.000502 | 0.000442 | -5.947284e-05 | -1.210446e-03 | -1.303 | -11.856237 | -10.511558 | -15.751934 | 8.272 | 6.969 | 17940.0 | 17820.0
/datum/controller/air_system/process_tiles_to_... | 0.000002 | 0.000001 | -5.498654e-07 | -5.498654e-07 | 0.000 | -28.121687 | -28.121687 | NaN | 0.000 | 0.000 | 17900.0 | 17788.0
/datum/controller/air_system/process_update_tiles | 0.000100 | 0.000200 | 1.000000e-04 | -1.000000e-03 | 0.000 | 100.000000 | -40.000000 | NaN | 0.000 | 0.000 | 40.0 | 40.0

name | avg_self_pre | avg_self_post | selfΔ | totalΔ | overtimeΔ | self% | total% | overtime% | over_pre | over_post | calls_pre | calls_post
---- | ------------ | ------------- | ----- | ------ | --------- | ----- | ------ | --------- | -------- | --------- | --------- | ---------- 
/datum/gas_mixture/New | 0.000002 | 6.648936e-07 | -9.629424e-07 | -9.629424e-07 | 0.000 | -59.154754 | -59.154754 | NaN | 0.000 | 0.000 | 9829.0 | 10528.0
/datum/gas_mixture/disposing | 0.000001 | 1.264468e-06 | -4.704728e-08 | -3.175021e-07 | -0.309 | -3.587246 | -9.399653 | -20.864281 | 1.481 | 1.172 | 2595471.0 | 2533081.0
/datum/gas_mixture/proc/check_both_then_temper... | 0.000005 | 4.108543e-06 | -1.051929e-06 | -1.047757e-06 | -0.386 | -20.384364 | -17.844611 | -39.630390 | 0.974 | 0.588 | 492203.0 | 475351.0
/datum/gas_mixture/proc/check_gas_mixture | 0.000016 | 9.972130e-06 | -5.892208e-06 | -5.981459e-06 | 0.005 | -37.141217 | -37.492874 | 18.518519 | 0.027 | 0.032 | 89635.0 | 91856.0
/datum/gas_mixture/proc/check_me_then_temperat... | 0.000003 | 3.037103e-06 | -1.545189e-07 | -1.477948e-07 | -0.260 | -4.841392 | -4.599204 | -6.804501 | 3.821 | 3.561 | 1326285.0 | 1364129.0
/datum/gas_mixture/proc/check_me_then_temperat... | 0.000004 | 2.910497e-06 | -6.106938e-07 | -6.857712e-07 | -0.748 | -17.343391 | -18.609038 | -15.334153 | 4.878 | 4.130 | 2268835.0 | 2295141.0
/datum/gas_mixture/proc/check_then_merge | 0.000009 | 3.553187e-06 | -5.281579e-06 | -1.601861e-05 | 0.000 | -59.781767 | -45.328340 | NaN | 0.000 | 0.000 | 4188.0 | 4503.0
/datum/gas_mixture/proc/check_then_remove | 0.000006 | 3.907098e-06 | -2.169088e-06 | -1.895837e-05 | 0.000 | -35.698185 | -50.386867 | NaN | 0.000 | 0.000 | 4279.0 | 4607.0
/datum/gas_mixture/proc/check_tile_graphic | 0.000001 | 7.610303e-07 | -4.707348e-07 | -4.736950e-07 | -15.329 | -38.216279 | -37.878563 | -44.484750 | 34.459 | 19.130 | 62828537.0 | 57207970.0
/datum/gas_mixture/proc/compare | 0.000002 | 2.113861e-06 | -4.438746e-08 | -5.748925e-08 | -0.762 | -2.056643 | -2.560904 | -20.934066 | 3.640 | 2.878 | 1846868.0 | 1509560.0
/datum/gas_mixture/proc/copy_from | 0.000004 | 3.000706e-06 | -9.448966e-07 | 7.600571e-07 | -0.017 | -23.948095 | 19.067582 | -14.406780 | 0.118 | 0.101 | 197435.0 | 185623.0
/datum/gas_mixture/proc/fire | 0.000006 | 5.291089e-06 | -5.630662e-07 | -7.389784e-07 | -8.586 | -9.618232 | -8.951932 | -23.209796 | 36.993 | 28.407 | 14037380.0 | 13413685.0
/datum/gas_mixture/proc/heat_capacity_archived... | 0.000002 | 1.861068e-06 | -5.302901e-07 | -5.355238e-07 | -6.638 | -22.175272 | -22.204568 | -31.389795 | 21.147 | 14.509 | 21503682.0 | 21275958.0
/datum/gas_mixture/proc/heat_capacity_full | 0.000002 | 1.895149e-06 | -4.692054e-07 | -4.695771e-07 | -15.055 | -19.844972 | -19.715846 | -26.194909 | 57.473 | 42.418 | 53021671.0 | 52710378.0
/datum/gas_mixture/proc/merge | 0.000009 | 7.968557e-06 | -1.463643e-06 | -2.381637e-06 | -3.156 | -15.517510 | -10.884788 | -29.611559 | 10.658 | 7.502 | 2595471.0 | 2533081.0
/datum/gas_mixture/proc/react | 0.000002 | 1.342198e-06 | -1.782961e-07 | -2.299947e-07 | -10.906 | -11.726194 | -6.511827 | -25.300422 | 43.106 | 32.200 | 65385335.0 | 59699100.0
/datum/gas_mixture/proc/remove | 0.000019 | 6.775355e-06 | -1.186948e-05 | -1.741938e-05 | 0.000 | -63.660964 | -55.115408 | NaN | 0.000 | 0.000 | 4398.0 | 4723.0
/datum/gas_mixture/proc/remove_ratio | 0.000008 | 7.037518e-06 | -1.423779e-06 | 2.963439e-07 | -3.032 | -16.826957 | 1.904457 | -31.645966 | 9.581 | 6.549 | 2567455.0 | 2503013.0
/datum/gas_mixture/proc/share | 0.000013 | 1.259755e-05 | -3.774167e-08 | 4.562834e-07 | -65.604 | -0.298700 | 2.619651 | -27.414617 | 239.303 | 173.699 | 42806123.0 | 33666775.0
/datum/gas_mixture/proc/subtract | 0.000005 | 2.138987e-06 | -3.067220e-06 | -3.067220e-06 | 0.000 | -58.914681 | -58.914681 | NaN | 0.000 | 0.000 | 19592.0 | 21038.0

File: code/modules/atmospherics/FEA_gas_mixture.dm [hunk 1]
```diff
@@ -115,7 +145,7 @@ What are the archived variables for?
 	if(length(src.trace_gases) > 0)
 		if(src.temperature > 900)
 			if(src.toxins > MINIMUM_HEAT_CAPACITY && src.carbon_dioxide > MINIMUM_HEAT_CAPACITY)
-				var/datum/gas/oxygen_agent_b/trace_gas = locate(/datum/gas/oxygen_agent_b/) in trace_gases
+				var/datum/gas/oxygen_agent_b/trace_gas = src.get_trace_gas_by_type(/datum/gas/oxygen_agent_b/)
```
Comment (Azrun) [inline code/modules/atmospherics/FEA_gas_mixture.dm:148]: See above

File: code/modules/atmospherics/FEA_gas_mixture.dm [hunk 2]
```diff
@@ -79,31 +81,59 @@ What are the archived variables for?
...
+
+	var/datum/gas/trace_gas = src.trace_gas_refs[type]
+	if(!trace_gas)
+		trace_gas = new type()
+		trace_gases += trace_gas
+		trace_gas_refs[type] = trace_gas
+	. = trace_gas
+
+/datum/gas_mixture/proc/get_trace_gas_by_type(type)
+	if(trace_gas_refs) . = src.trace_gas_refs[type]
```
Comment (Azrun) [inline code/modules/atmospherics/FEA_gas_mixture.dm:120]: DOCS DOCS DOCS

File: _std/macros/atmos.dm [hunk 3]
```diff
@@ -17,6 +17,9 @@
 /// Moles in a standard cell after which plasma is visible
 #define MOLES_PLASMA_VISIBLE	2
 
+#define GAS_IMG_PLASMA_BIT (1<<0)
+#define GAS_IMG_N2O_BIT	(1<<1)
```
Comment (Azrun) [inline _std/macros/atmos.dm:21]: DOCS

File: code/modules/atmospherics/FEA_gas_mixture.dm [hunk 4]
```diff
@@ -42,13 +42,15 @@ What are the archived variables for?
 	var/graphic
 	var/tmp/graphic_archived // intentionally NOT using ARCHIVED() because graphic archiving is actually important and shouldn't be turned off
 	var/list/datum/gas/trace_gases
+	var/list/trace_gas_refs // mapping of type=gas to leverage hashing for locate and avoid O(n^2)
```
Comment (Azrun) [inline code/modules/atmospherics/FEA_gas_mixture.dm:45]: O(n^2) is only applicable when comparing mixes, be specific or omit specifics.

**Review by ZeWaka — APPROVED** [review]
reviewed all.

**Review by pali6 — APPROVED** [review]
Didn't test it locally but code looks good!

**Pull request merged**